test/name_collisions_test.rb

```diff
@@ -0,0 +1,160 @@
+require 'test_helper'
+$VERBOSE=nil #otherwise we get lots of C:/gems/surrounded/lib/surrounded/context/name_collision_detector.rb:25: warning: instance variable @handler not initialized
+
+##
+# The problem is as follows: When an object already contains a method
+# that has the same name as an actor in the context, then the object
+# defaults to its own method rather than calling the actor.
+# This is simulated in the two classes below, where HasNameCollision has
+# an empty attribute called will_collide (which is always nil). In the following
+# context, when the method collide is called on the actor will_collide, the
+# actor is ignored and the HasNameCollision#will_collide is called instead,
+# returning nil.
+##
+
+class HasNameCollision
+  include Surrounded
+
+  attr_accessor :will_collide # should always return nil
+
+  def assert_has_role
+    true
+  end
+
+  def will_collide=(val); end
+
+end
+
+class ShouldCollide
+  include Surrounded
+  def collide
+    return 'Method called in ShouldCollide'
+  end
+
+end
+
+class ContextOverridesName
+  extend Surrounded::Context
+
+  # The specific behaviour we want to test is that when a role has a method that
+  # has the same name as another role, then when that method is called strange
+  # things happen.
+  keyword_initialize :base, :will_collide
+  on_name_collision :nothing
+
+  trigger :check_setup do
+    base.assert_has_role
+  end
+
+  trigger :induce_collision do
+    base.name_collision
+  end
+
+  role :base do
+    def name_collision
+      will_collide.collide
+    end
+
+    def assert_has_role
+      true
+    end
+  end
+end
+
+class ContextWithMultipleCollisions
+  extend Surrounded::Context
+
+    on_name_collision :warn
+
+    keyword_initialize :first, :second, :third
+
+end
+
+class First
+  def second
+  end
+
+  def third
+  end
+end
+
+class Second
+
+  def first
+  end
+
+  def third
+  end
+end
+
+class Third
+
+  def first
+  end
+
+  def second
+  end
+
+end
+
+
+# Just check that the basic context is properly set up first
+describe Surrounded::Context, 'context correctly set up' do
+
+  let(:has_collision){
+    ContextOverridesName.new(base: HasNameCollision.new, will_collide: ShouldCollide.new)
+  }
+
+  let(:multiple_collisions){
+    ContextWithMultipleCollisions.new(first: First.new, second: Second.new, third: Third.new)
+  }
+
+  after do
+    ContextOverridesName.instance_eval{
+      on_name_collision :nothing
+    }
+  end
+
+  it 'is surrounded' do
+    assert has_collision.check_setup
```

Comment:
I think the collision should happen on `ContextOverridesName.new` not on calling a method on that object.
In other words, I'd expect the test to be something like `assert_raises(WhateverException){ ContextOverridesName.new }`